### PR TITLE
CUDA: Add two-argument implementation of round()

### LINKED ISSUE #4668: Round segfaults in cuda jitted function if passed ndigits optional parameter
## Reporting a bug

- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

Reproducer:

```python
import numpy as np
from numba import cuda

@cuda.jit
def gpu_round(in_array, out_array, digits=1):
    i = cuda.grid(1)
    if i < in_array.size:
        out_array[i] = round(in_array[i], digits)

test_in = cuda.to_device(np.array([0.123, 0.4567, 0.78, 0.9]))
test_out = cuda.to_device(np.array([0., 0., 0., 0.]))

gpu_round(test_in, test_out, 2)
print(test_out.copy_to_host())
```

<details>
<summary>cuda-gdb backtrace:</summary>

```
(cuda-gdb) bt
#0  0x00007fff29bb0d85 in nvvmVerifyProgram () from /home/nfs/kkraus/miniconda3/envs/cudf_dev/lib/libnvvm.so.3.3.0
#1  0x00007fff29bb8b84 in nvvmVerifyProgram () from /home/nfs/kkraus/miniconda3/envs/cudf_dev/lib/libnvvm.so.3.3.0
#2  0x00007fff298a69ce in nvvmVerifyProgram () from /home/nfs/kkraus/miniconda3/envs/cudf_dev/lib/libnvvm.so.3.3.0
#3  0x00007fff2985ba70 in nvvmVerifyProgram () from /home/nfs/kkraus/miniconda3/envs/cudf_dev/lib/libnvvm.so.3.3.0
#4  0x00007fff2985d03b in nvvmVerifyProgram () from /home/nfs/kkraus/miniconda3/envs/cudf_dev/lib/libnvvm.so.3.3.0
#5  0x00007fff2986774b in nvvmVerifyProgram () from /home/nfs/kkraus/miniconda3/envs/cudf_dev/lib/libnvvm.so.3.3.0
#6  0x00007fff2986a81f in nvvmVerifyProgram () from /home/nfs/kkraus/miniconda3/envs/cudf_dev/lib/libnvvm.so.3.3.0
#7  0x00007fff2977dbc5 in nvvmVerifyProgram () from /home/nfs/kkraus/miniconda3/envs/cudf_dev/lib/libnvvm.so.3.3.0
#8  0x00007fff29780014 in nvvmVerifyProgram () from /home/nfs/kkraus/miniconda3/envs/cudf_dev/lib/libnvvm.so.3.3.0
#9  0x00007fff29781550 in nvvmVerifyProgram () from /home/nfs/kkraus/miniconda3/envs/cudf_dev/lib/libnvvm.so.3.3.0
#10 0x00007fff29a95ae8 in nvvmVerifyProgram () from /home/nfs/kkraus/miniconda3/envs/cudf_dev/lib/libnvvm.so.3.3.0
#11 0x00007fff29a95cdb in nvvmVerifyProgram () from /home/nfs/kkraus/miniconda3/envs/cudf_dev/lib/libnvvm.so.3.3.0
#12 0x00007fff29a94f6f in nvvmVerifyProgram () from /home/nfs/kkraus/miniconda3/envs/cudf_dev/lib/libnvvm.so.3.3.0
#13 0x00007fff295e5202 in ?? () from /home/nfs/kkraus/miniconda3/envs/cudf_dev/lib/libnvvm.so.3.3.0
#14 0x00007fff295f7b1e in nvvmCompileProgram () from /home/nfs/kkraus/miniconda3/envs/cudf_dev/lib/libnvvm.so.3.3.0
#15 0x00007ffff7e4d630 in ffi_call_unix64 () from /home/nfs/kkraus/miniconda3/envs/cudf_dev/lib/python3.6/lib-dynload/../../libffi.so.6
#16 0x00007ffff7e4cfed in ffi_call () from /home/nfs/kkraus/miniconda3/envs/cudf_dev/lib/python3.6/lib-dynload/../../libffi.so.6
#17 0x00007fff2e9da00e in _call_function_pointer (argcount=3, resmem=0x7fffffffc310, restype=<optimized out>, atypes=0x7fffffffc2b0,
    avalues=0x7fffffffc2e0, pProc=0x7fff295f6dd0 <nvvmCompileProgram>, flags=4353)
    at /usr/local/src/conda/python-3.6.7/Modules/_ctypes/callproc.c:809
#18 _ctypes_callproc () at /usr/local/src/conda/python-3.6.7/Modules/_ctypes/callproc.c:1166
#19 0x00007fff2e9daa45 in PyCFuncPtr_call () at /usr/local/src/conda/python-3.6.7/Modules/_ctypes/_ctypes.c:3962
#20 0x00005555556677bb in _PyObject_FastCallDict ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Objects/abstract.c:2331
#21 0x00005555556f75be in call_function () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4861
#22 0x000055555571a73a in _PyEval_EvalFrameDefault ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:3335
#23 0x00005555556c2056 in _PyEval_EvalCodeWithName ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4166
#24 0x00005555556c376a in _PyFunction_FastCallDict ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:5070
#25 0x000055555566799f in _PyObject_FastCallDict ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Objects/abstract.c:2310
#26 0x000055555566c613 in _PyObject_Call_Prepend ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Objects/abstract.c:2373
#27 0x00005555556673de in PyObject_Call () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Objects/abstract.c:2261
#28 0x000055555571be56 in do_call_core (kwdict=0x7ffff7f18510, callargs=0x7ffff7fa0048, func=0x7ffff7f92988)
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:5106
#29 _PyEval_EvalFrameDefault () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:3404
#30 0x00005555556c6749 in _PyEval_EvalCodeWithName (qualname=0x0, name=<optimized out>, closure=0x0, kwdefs=0x0, defcount=0, defs=0x0,
    kwstep=2, kwcount=<optimized out>, kwargs=0x7fffef42fae8, kwnames=0x7fffef42fae0, argcount=<optimized out>, args=0x555556f197b0,
    locals=0x0, globals=<optimized out>, _co=0x7ffff39355d0)
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4166
#31 PyEval_EvalCodeEx () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4187
#32 0x00005555556c7676 in function_call ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Objects/funcobject.c:604
#33 0x00005555556673de in PyObject_Call () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Objects/abstract.c:2261
#34 0x000055555571be56 in do_call_core (kwdict=0x7fffee6fc510, callargs=0x7fffee6d0e10, func=0x7ffff3947e18)
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:5106
#35 _PyEval_EvalFrameDefault () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:3404
#36 0x00005555556c2e4b in _PyFunction_FastCall (globals=<optimized out>, nargs=1, args=<optimized out>, co=<optimized out>)
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4919
#37 fast_function () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4954
#38 0x00005555556f7545 in call_function () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4858
#39 0x000055555571a73a in _PyEval_EvalFrameDefault ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:3335
#40 0x00005555556c2e4b in _PyFunction_FastCall (globals=<optimized out>, nargs=1, args=<optimized out>, co=<optimized out>)
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4919
#41 fast_function () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4954
#42 0x00005555556f7545 in call_function () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4858
#43 0x000055555571a73a in _PyEval_EvalFrameDefault ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:3335
#44 0x00005555556c2e4b in _PyFunction_FastCall (globals=<optimized out>, nargs=1, args=<optimized out>, co=<optimized out>)
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4919
#45 fast_function () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4954
#46 0x00005555556f7545 in call_function () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4858
#47 0x000055555571a73a in _PyEval_EvalFrameDefault ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:3335
#48 0x00005555556c2e4b in _PyFunction_FastCall (globals=<optimized out>, nargs=2, args=<optimized out>, co=<optimized out>)
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4919
#49 fast_function () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4954
#50 0x00005555556f7545 in call_function () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4858
#51 0x000055555571a73a in _PyEval_EvalFrameDefault ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:3335
#52 0x00005555556c248e in _PyEval_EvalCodeWithName ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4166
#53 0x00005555556c354c in _PyFunction_FastCallDict ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:5070
#54 0x000055555566799f in _PyObject_FastCallDict ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Objects/abstract.c:2310
#55 0x000055555566c613 in _PyObject_Call_Prepend ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Objects/abstract.c:2373
#56 0x00005555556673de in PyObject_Call () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Objects/abstract.c:2261
#57 0x000055555571be56 in do_call_core (kwdict=0x0, callargs=0x7ffff34c5708, func=0x7ffff7ee0d08)
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:5106
#58 _PyEval_EvalFrameDefault () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:3404
#59 0x00005555556c2056 in _PyEval_EvalCodeWithName ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4166
#60 0x00005555556c354c in _PyFunction_FastCallDict ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:5070
#61 0x000055555566799f in _PyObject_FastCallDict ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Objects/abstract.c:2310
#62 0x000055555566c613 in _PyObject_Call_Prepend ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Objects/abstract.c:2373
#63 0x00005555556673de in PyObject_Call () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Objects/abstract.c:2261
#64 0x00005555556cd397 in slot_tp_call ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Objects/typeobject.c:6207
#65 0x00005555556677bb in _PyObject_FastCallDict ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Objects/abstract.c:2331
#66 0x00005555556f75be in call_function () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4861
#67 0x000055555571a73a in _PyEval_EvalFrameDefault ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:3335
#68 0x00005555556c6749 in _PyEval_EvalCodeWithName (qualname=0x0, name=<optimized out>, closure=0x0, kwdefs=0x0, defcount=0, defs=0x0,
    kwstep=2, kwcount=<optimized out>, kwargs=0x0, kwnames=0x0, argcount=0, args=0x0, locals=0x7ffff7f581f8, globals=0x7ffff7f581f8,
    _co=0x7ffff65c2540) at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4166
#69 PyEval_EvalCodeEx () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:4187
#70 0x00005555556c74ec in PyEval_EvalCode (co=co@entry=0x7ffff65c2540, globals=globals@entry=0x7ffff7f581f8,
    locals=locals@entry=0x7ffff7f581f8) at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/ceval.c:731
#71 0x000055555576d584 in run_mod () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/pythonrun.c:1025
#72 0x000055555576d981 in PyRun_FileExFlags ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/pythonrun.c:978
#73 0x000055555576db83 in PyRun_SimpleFileExFlags ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/pythonrun.c:419
#74 0x000055555576dc8d in PyRun_AnyFileExFlags ()
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Python/pythonrun.c:81
#75 0x00005555557716f7 in run_file (p_cf=0x7fffffffdf2c, filename=0x5555558ac6e0 L"test.py", fp=0x555555900e40)
    at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Modules/main.c:340
#76 Py_Main () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Modules/main.c:810
#77 0x000055555563832e in main () at /home/conda/feedstock_root/build_artifacts/python_1562031814771/work/Programs/python.c:69
```

</details>

### LINKED ISSUE #5708: libnvvm segfault compiling this kernel
## Reporting a bug

- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

From gitter https://gitter.im/numba/numba?at=5ebef15eeb9b6f316225be0e

Reproducer:

```python
from numba import cuda
import numba

# Compiling this will segfault inside libnvvm
@cuda.jit(argtypes=[numba.float64[:]])
def kernel(arr):
    arr[cuda.threadIdx.x] = cuda.threadIdx.x / 100.0 + 2
    a_local = 0
    if round(arr[cuda.threadIdx.x], 2) > 0.09:
        a_local += 1
        arr[cuda.threadIdx.x] = a_local
```

lldb shows:

```
Process 50672 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
    frame #0: 0x000000011809b3ef libnvvm.3.2.0.dylib`llvm::NVPTXTargetLowering::getArgumentAlignment(llvm::SDValue, llvm::ImmutableCallSite const*, llvm::Type*, unsigned int) const + 31
libnvvm.3.2.0.dylib`llvm::NVPTXTargetLowering::getArgumentAlignment:
->  0x11809b3ef <+31>: movq   (%rcx), %rax
    0x11809b3f2 <+34>: movq   %rax, %r13
    0x11809b3f5 <+37>: andq   $-0x4, %r13
    0x11809b3f9 <+41>: leaq   -0x18(%r13), %rcx
```
Happened on windows with `cudatoolkit 10.2` and MBP with `cudatoolkit 9.0`.

### PR DESCRIPTION
**Pull request by gmarkall**
This is based on the `cpython.builtins.round_impl_binary()` function, but without using the `llvm.rint` intrinsic - instead it lowers using the already-implemented `__nv_rint()` intrinsic.

Fixes #5708.
Fixes #4668.

### CODE REVIEW

File: numba/cuda/cudaimpl.py [hunk 1]
```diff
@@ -468,6 +468,35 @@ def ptx_round(context, builder, sig, args):
...
+    def round_ndigits(x, ndigits):
+        if math.isinf(x) or math.isnan(x):
+            return x
+
+        if ndigits >= 0:
+            if ndigits > 22:
+                # pow1 and pow2 are each safe from overflow, but
+                # pow1*pow2 ~= pow(10.0, ndigits) might overflow.
+                pow1 = 10.0 ** (ndigits - 22)
+                pow2 = 1e22
```
Comment (stuartarchibald) [inline numba/cuda/cudaimpl.py:483]: Is this branch hit in the tests?

File: numba/cuda/cudaimpl.py [hunk 2]
```diff
@@ -468,6 +468,35 @@ def ptx_round(context, builder, sig, args):
...
+            else:
+                pow1 = 10.0 ** ndigits
+                pow2 = 1.0
+            y = (x * pow1) * pow2
+            if math.isinf(y):
+                return x
```
Comment (stuartarchibald) [inline numba/cuda/cudaimpl.py:489]: Same with this branch?

File: numba/cuda/cudaimpl.py [hunk 3]
```diff
@@ -468,6 +468,35 @@ def ptx_round(context, builder, sig, args):
     ])
 
 
+@lower(round, types.f4, types.Integer)
+@lower(round, types.f8, types.Integer)
+def round_to_impl(context, builder, sig, args):
+    def round_ndigits(x, ndigits):
+        if math.isinf(x) or math.isnan(x):
+            return x
```
Comment (stuartarchibald) [inline numba/cuda/cudaimpl.py:476]: I this branch hit in the tests?

**Comment by gmarkall** [comment]
Thanks for the review @stuartarchibald - having done more extensive testing, I've now found that the implementation I copied from, in the CPU target, has at least two issues - it doesn't correctly round when the value is exactly halfway between two values to round to, and it has quite limited range for single precision values. Unfortunately I'd assumed that the CPU target code was already working correctly!

I'm not going to be able to sort this out for 0.50 (I imagine) so I'll change the milestone. I'll push a couple of commits of my current state, but this needs a little more work to be ready.

**Comment by gmarkall** [comment]
I'm not going to be able to sort this out for 0.51RC, so I've bumped it to 0.52RC.

**Comment by gmarkall** [comment]
I've picked this up and moved it back to the 0.52RC milestone because it fixes a couple of issues that have been waiting since 0.49.

It turned out that the test fail in `test_round_to_f4_halfway` was because I was testing with a value that was not actually halfway between two values due to limited precision when using `float32` - I've changed the value now to something that can be more precisely represented, and the test now passes - I've added a test with `float64` with the original value I was testing with, which does pass also.

I believe all branches are now getting hit by the tests that are added here.

**Comment by gmarkall** [comment]
@stuartarchibald I think this is now ready for another round of review.

File: numba/cuda/cudaimpl.py [hunk 4]
```diff
@@ -468,6 +468,48 @@ def ptx_round(context, builder, sig, args):
     ])
 
 
+# This rounding implementation follows the algorithm used in the "fallback
+# version" of double_round in CPython 3.7.5's Objects/floatobject.c.
```
Comment (stuartarchibald) [inline numba/cuda/cudaimpl.py:472]: Please add an exact reference with SHA to the implementation. Example:
https://github.com/numba/numba/blob/626b40eddc9a29a1e89bfb6b8706e4f390cfcc4a/numba/cpython/hashing.py#L30-L31

File: numba/cuda/tests/cudapy/test_intrinsics.py [hunk 5]
```diff
@@ -394,6 +399,70 @@ def test_round_f8(self):
...
+            13
+        )
+        for val, ndigits in itertools.product(vals, digits):
+            with self.subTest(val=val, ndigits=ndigits):
+                compiled[1, 1](ary, val, ndigits)
+                self.assertPreciseEqual(ary[0], round(val, ndigits),
+                                        prec='single')
+
+    def test_round_to_f4_halfway(self):
+        # Presently not passing
```
Comment (stuartarchibald) [inline numba/cuda/tests/cudapy/test_intrinsics.py:423]: This is passing?!

File: numba/cuda/tests/cudapy/test_intrinsics.py [hunk 6]
```diff
@@ -394,6 +399,70 @@ def test_round_f8(self):
             compiled[1, 1](ary, i)
             self.assertEquals(ary[0], round(i))
 
+    def test_round_to_f4(self):
```
Comment (stuartarchibald) [inline numba/cuda/tests/cudapy/test_intrinsics.py:402]: I'm not sure the unit tests trigger this branch https://github.com/numba/numba/pull/5741/files#diff-a3ad32c82e29f57220d54a522266e48cR491-R492

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the fixes. I've checked the alg change against the CPython impl and it now appears to be an identical algorithm to the noted original. Just a few minor things to look at else good to go. Thanks!

File: numba/cuda/tests/cudapy/test_intrinsics.py (hunk 5 above)
Comment (gmarkall) [reply numba/cuda/tests/cudapy/test_intrinsics.py:423]: Accidentally left in this comment from an earlier commit - it is indeed passing now :-)

File: numba/cuda/tests/cudapy/test_intrinsics.py (hunk 6 above)
Comment (gmarkall) [reply numba/cuda/tests/cudapy/test_intrinsics.py:404]: I've added tests to trigger this branch. I tested that they hit the branch by applying:

```diff
diff --git a/numba/cuda/cudaimpl.py b/numba/cuda/cudaimpl.py
index 7e6cb1832..6eb425f10 100644
--- a/numba/cuda/cudaimpl.py
+++ b/numba/cuda/cudaimpl.py
@@ -490,6 +490,7 @@ def round_to_impl(context, builder, sig, args):
                 pow2 = 1.0
             y = (x * pow1) * pow2
             if math.isinf(y):
+                print(11111)
                 return x
 
         else:
```

and checking that the test cases printed out `11111`.

**Comment by gmarkall** [comment]
> @stuartarchibald Thanks for the review! Everything should be resolved now.

... apart from the new failure with the simulator - looking into it now - should be straightforward to resolve.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the fixes!

**Comment by stuartarchibald** [comment]
Buildfarm ID: `numba_smoketest_cuda_110`

**Comment by stuartarchibald** [comment]
> Buildfarm ID: `numba_smoketest_cuda_110`

passed.

**Pull request merged**